packages/core/src/electron-main/messaging/electron-main-messaging-service.ts

```diff
@@ -0,0 +1,120 @@
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import { inject, injectable, named } from 'inversify';
+import { createWebSocketConnection } from 'vscode-ws-jsonrpc/lib/socket/connection';
+import { ContributionProvider } from '../../common/contribution-provider';
+import { WebSocketChannel } from '../../common/messaging/web-socket-channel';
+import { ConsoleLogger } from '../../node/messaging/logger';
+import { ElectronMainConnectionHandler, ElectronBackendMessage, ElectronBackendConnectionPipe } from '../../electron-common/messaging/electron-backend-connection-handler';
+import { MessagingContribution } from '../../node/messaging/messaging-contribution';
+import { ChildProcess } from 'child_process';
+
+export interface ElectronMainConnectionOptions {
+}
+
+/**
+ * This component replicates the role filled by `MessagingContribution` but for connecting the backend with the electron process.
+ * It is based on process communication using the created backend process.
+ * Alternatively it uses a simple message pipe object when running in `--no-cluster` mode.
+ *
+ * This component allows communication between server process (backend) and electron main process.
+ */
+@injectable()
+export class ElectronMainMessagingService {
+
+    @inject(ContributionProvider) @named(ElectronMainConnectionHandler)
+    protected readonly connectionHandlers: ContributionProvider<ElectronMainConnectionHandler>;
+
+    protected readonly channelHandlers = new MessagingContribution.ConnectionHandlers<WebSocketChannel>();
+    protected readonly channels = new Map<number, WebSocketChannel>();
+    protected backendProcess?: ChildProcess;
+
+    start(backendProcess?: ChildProcess): void {
+        if (backendProcess) {
+            this.backendProcess = backendProcess;
+            this.backendProcess.on('message', message => {
+                if (ElectronBackendMessage.is(message)) {
+                    this.handleMessage(ElectronBackendMessage.get(message));
+                }
+            });
+            this.backendProcess.on('exit', () => {
+                this.closeChannels();
+            });
+        } else {
+            ElectronBackendConnectionPipe.onMessage('electron', message => {
+                this.handleMessage(message);
+            });
+        }
+        for (const connectionHandler of this.connectionHandlers.getContributions()) {
+            this.channelHandlers.push(connectionHandler.path, (params, channel) => {
+                const connection = createWebSocketConnection(channel, new ConsoleLogger());
+                connectionHandler.onConnection(connection);
+            });
+        }
+    }
+
+    protected closeChannels(): void {
+        for (const channel of Array.from(this.channels.values())) {
```

Comment:
```suggestion
        for (const channel of this.channels.values()) {
```

This should already be iterable.